### PR TITLE
Prevent ForOfStatement from printing the forbidden sequence "for ( async of"

### LINKED ISSUE #13207: [Bug]: ForOfStatement prints the forbidden token sequence "for ( async of"
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
for ((async) of []);
for (\u0061sync of []);
```

[Babel REPL link](https://babel.dev/repl#?browsers=&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=GYewTgBAFFCGDOBPAdgYwJQRMCBtAuugNwBQokUAOgK4AMtAbAIxJpY4HFA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.13.17&externalPlugins=)

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

Expected: The `await` identifier is parenthesized or otherwise escaped.
```js
for ((async) of []);

for ((async) of []);
```

Actual: The generator prints illegal syntax:
```js
for (async of []);

for (async of []);
```

### Environment

- Current Babel main (10f4d08efbba49b4085f0756b59c835f97783fa7)

### Possible solution

Detect when a `ForOfStatement`'s `left` node is an identifier named `async`, and print extra parentheses around the identifier.

### Additional context

https://github.com/tc39/ecma262/pull/2256

### PR DESCRIPTION
**Pull request by Zalathar**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #13207 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | No
| License                  | MIT

If we detect that a `ForOfStatement`'s left-hand-side is an identifier named `async`, print extra parentheses around the identifier to avoid this edge-case in the ECMAScript grammar.

### CODE REVIEW

**Comment by Zalathar** [comment]
After writing this I remembered that there's a whole subsystem dedicated to “does this node need parentheses in light of its parent”. Not sure if it makes more sense to to use that instead, or to stick with this ad-hoc code.

**Comment by Zalathar** [comment]
The alternative would be something like this in `parentheses.ts`:

```typescript
export function Identifier(node: t.Identifier, parent: t.Node): boolean {
  return (
    node.name === "async" && t.isForOfStatement(parent) && node === parent.left
  );
}


```

**Comment by existentialism** [comment]
> The alternative would be something like this in `parentheses.ts`:
> 
> ```ts
> export function Identifier(node: t.Identifier, parent: t.Node): boolean {
>   return (
>     node.name === "async" && t.isForOfStatement(parent) && node === parent.left
>   );
> }
> ```

:+1: from me

File: packages/babel-generator/src/generators/statements.ts [hunk 1]
```diff
@@ -82,16 +82,26 @@ export function WhileStatement(this: Printer, node: t.WhileStatement) {
...
     }
     this.token("(");
+
+    // ECMAScript specifically forbids a for-of loop from starting with the
+    // token sequence "for ( async of", because it would be ambiguous with
+    // "for (async of => {};;)", so we need to add extra parentheses.
+    const leftNeedsParens =
+      op === "of" && t.isIdentifier(node.left) && node.left.name === "async";
+
+    if (leftNeedsParens) this.token("(");
```
Comment (JLHwung) [inline packages/babel-generator/src/generators/statements.ts:101]: We have `packages/babel-generator/src/node/parentheses.ts` for determining whether an AST node should be wrapped in parentheses.

You can add a new `Identifier` handler and check if the `parent` is a forOfStatement and the node comes from left.

**Review by JLHwung — APPROVED** [review]

**Comment by JLHwung** [comment]
@existentialism Oops I mentally think of your 👍 as ✅.

**Comment by existentialism** [comment]
All good, looks great thanks @Zalathar!

**Comment by thorn0** [comment]
Looks like this PR doesn't take into account that in `for await (async of []);` the parens are not needed.

**Comment by Zalathar** [comment]
> Looks like this PR doesn't take into account that in `for await (async of []);` the parens are not needed.

Ah, nice catch.

Thankfully the output still happens to be *correct*, so it's not a disaster, but this was an oversight and at least deserves explicit tests and comments.

That said, some quick testing reveals that Babel itself doesn't know how to parse `for await (async of []);`, which is more concerning. So the extra parens turn out to be an unintended mitigation for a parser bug, as well as a hand-hold for any unfortunate humans who have to read this weird code.

(TODO: File and fix the parser bug!)

So once the parser bug is fixed, I'm inclined to leave the unnecessary parens as-is, or *maybe* remove them in compact mode only.

**Pull request merged**